### ISSUE TITLE
6.3 Beta performance regression on windows?

### ISSUE THREAD

**Issue by mpharrigan** [pandegroup/openmm#969]
I built openmm 6.2 and 6.3-beta1 on windows (x64, release, msvc2013, cuda 6.5) and built fahbench (https://github.com/pandegroup/fahbench) against each of them. This runs dhfr for a couple thousand steps and reports performance.

These are my numbers on a gtx 570 on the opencl platform

precision | 6.2 | 6.3
-----------|-----|------
single | 23.4 | 2.0
double | 7.18 | error

**Comment by mpharrigan** [comment]
Is there something I could have messed up with the build? Have any of the cmake options changed?

**Comment by informatorius** [comment]
I can confirm FahBench on AMD R9 280 with OpenMM 6.2 it was 38 ns/day and with 6.3-beta1 it is 7 ns/day

**Comment by peastman** [comment]
Here's what I get on a 7970:

Linux: 49 ns/day
Windows: 23 ns/day

This is on the exact same computer, just booting into different OSes.  So it's clearly something Windows specific.  Although the slowdown is much less than what you're seeing.  On the other hand, this is with the OpenMM benchmarking script.  I don't know much about FAHBench, and I can't rule out that it's doing something strange.

Anyway, I'll investigate.

**Comment by informatorius** [comment]
What is your OpenMM benchmarking script values in ns/day in OpenMM 6.2 compared to 6.3beta1?
How can i run the OpenMM benchmarking script myself?

**Comment by peastman** [comment]
It's included in the examples directory.  Try this:

```
python benchmark.py --platform=OpenCL --test=pme --seconds=10
```

That will give you a quick but still reasonably accurate number.  Increasing the number of seconds will make it more accurate.

**Comment by peastman** [comment]
Recompiling against the most recent AMD OpenCL library brings the Windows number up to 46 ns/day.  I have no idea why that should make a difference, but apparently it does!

But @mpharrigan is using NVIDIA, so that's clearly something different.  I'll investigate that next.

**Comment by peastman** [comment]
On NVIDIA, there's something very strange going on.  I get roughly twice the performance running under Linux as Windows.  That's true with both CUDA and OpenCL.  Using nvvp to profile it, I see that the GPU spends a large part of its time sitting idle.  It executes a few time steps (the number varies but is usually about six), then does nothing for a few milliseconds, then runs some more time steps.  I've been unable so far to determine why.  I'll post an update if I get more information.

**Comment by swails** [comment]
> I see that the GPU spends a large part of its time sitting idle.

Perhaps that's why they call it Windoze?

**Comment by mpharrigan** [comment]
Thoughts:

 - Could this be because of the new threading library on windows?
 - Why don't we see the performance regression on @rmcgibbo 's airspeed velocity thing
 - Did you try upgrading nvidia drivers (it worked for amd, lol)
 - I should try upgrading my nvidia drivers

**Comment by peastman** [comment]
> Could this be because of the new threading library on windows?

No.  The only case where the CUDA and OpenCL platforms use multiple threads is when running on multiple GPUs.

> Why don't we see the performance regression on @rmcgibbo 's airspeed velocity thing

Does he have benchmarks on identical hardware under Linux and Windows?  If so, what do they show?  (I can't check, because the page seems to be broken right now.)

> Did you try upgrading nvidia drivers (it worked for amd, lol)

I'm using very recent drivers.

> I should try upgrading my nvidia drivers

Go for it!

**Comment by informatorius** [comment]
Maybe someone can try to build with 6.2 + some changesets between 6.3b1 to see which changeset brought in the performance regression on Windows ...

**Comment by rmcgibbo** [comment]
I did this, and didn't really track a systematic change.

http://rmcgibbo.github.io/openmm-bench/#pme.track_opencl

**Comment by peastman** [comment]
The page you linked to is broken.

**Comment by informatorius** [comment]
The page is working for me :-)
So maybe it is a FahBench / Folding@Home only issue.
Someone should build OpenMM 6.2 to 6.3b1 changesets and run with FahBench to see which changeset broke FahBench.

**Comment by peastman** [comment]
Here's what I get:

> Can not determine continental origin of swallow.
> One or more external dependencies of airspeed velocity failed to load.
> Make sure you have an active internet connection the first time you load airspeed velocity.

**Comment by rmcgibbo** [comment]
@peastman: weird. works for me (laptop / wifi)

**Comment by mpharrigan** [comment]
@rmcgibbo : that's amd, though. right?

**Comment by rmcgibbo** [comment]
right. is this issue only nvidia?

**Comment by peastman** [comment]
I suspect it's because I have NoScript installed.  It provides a menu so I can selectively unblock scripts coming from different domains.  Unfortunately, Airspeed Velocity doesn't let me do that.  As soon as it discovers anything can't load, it immediately redirects me to an error page, so I never have a chance to see what's being blocked and unblock it.

**Comment by peastman** [comment]
We're talking about two different things: performance problems with Nvidia, and Robert's benchmarking page. :)

**Comment by peastman** [comment]
There was a different issue with AMD, but that's resolved by compiling against the AMD OpenCL library.

**Comment by informatorius** [comment]
This issue is Nvidia and AMD. I could reproduce this problem on my AMD R9 280

**Comment by peastman** [comment]
No, there are two unrelated problems.  The one involving AMD has been resolved.  The one involving Nvidia has not, and looks like it may be caused by a driver bug.

**Comment by informatorius** [comment]
1) OK, so https://github.com/pandegroup/openmm/issues/949 is AMD issue and is solved by building against AMD APP SDK.
2) This issue #969 was first found with Nvidia GPU then i found AMD GPU also has issue #969
=> So we have issue 1) #949 for AMD only and 2) #969 for Nvidia and AMD
=> Should i open a new issue for #969 with AMD GPU or will it be tracked here?

**Comment by peastman** [comment]
Let's just limit this issue to discussing Nvidia, and use #949 for AMD.

**Comment by informatorius** [comment]
But this issue #969 is about OpenMM 6.3b1 and is different from issue #949 OpenMM 6.2
1) Issue #949 with OpenMM 6.2 on AMD GPU can be solved be compiling against AMD APP SDK.
2) Issue #969 with OpenMM 6.3b1 on AMD GPU still fails when compiling against AMD APP SDK.
When we discuss both issues for AMD in issue 1) #949 then we mix the two unrelated problems.

**Comment by peastman** [comment]
The beta of 6.3 was compiled against the Nvidia OpenCL.  When the code is compiled again the AMD OpenCL, the performance problem goes away.  The official 6.3 release will be compiled against AMD OpenCL.  I don't believe there's any other outstanding issue related to AMD.

**Comment by jchodera** [comment]
> The official 6.3 release will be compiled against AMD OpenCL. I don't believe there's any other outstanding issue related to AMD.

Can you clarify if I need to upgrade the Jenkins build machine scripts to do this for `linux` and `osx`, or if this is just for your `win` build?

**Comment by informatorius** [comment]
@peastman: It's not working on my side, this is why i keep this under investigation. 
What are your settings to compile OpenMM 6.3b1 with AMD OpenCL on Windows and use it with FahBench?
I use Windows 7 64bit, Visual Studio 2013 Update 4, cmake 3.2.2, Python 3.4.3, AMD APP SDK 2.9-1
1) Build OpenMM and FahBench 2.1.2 in 64bit release
2) Run FahBench on AMD R9 280
=> OpenMM 6.2	= 38 ns/day
=> OpenMM 6.3b1 =  7 ns/day

**Comment by peastman** [comment]
@jchodera The Linux and OS X builds *should* have been building against AMD OpenCL all along - otherwise, they can get link errors when running on an AMD GPU.  (But if you can confirm they really are, that would be good.)  It's only the Windows version that I built by hand that accidentally got built against the Nvidia one.

@informatorius The import thing is the values for the following CMake variables: OPENCL_INCLUDE_DIR and OPENCL_LIBRARY.  Make sure those point to the AMD versions.

**Comment by jchodera** [comment]
> @jchodera The Linux and OS X builds should have been building against AMD OpenCL all along - otherwise, they can get link errors when running on an AMD GPU. (But if you can confirm they really are, that would be good.) It's only the Windows version that I built by hand that accidentally got built against the Nvidia one.

I can confirm they are *not* using the AMD OpenCL.

The `linux` VM that Jenkins uses is here:
https://github.com/omnia-md/virtual-machines

According to the [history](https://github.com/omnia-md/virtual-machines/commits/master/linux/setup_centos_vm.sh), I expunged the AMD APP SDK on 3 Nov 2014 because it was giving us trouble on automating vagrant provisioning.

I'll note that I tried to capture the exact build environment you specified for these packages, and they did not include building against the AMD APP SDK:
https://github.com/pandegroup/openmm/wiki/OpenMM-build-platforms-and-environments

The Jenkins `osx` machine does not have any AMD SDK installed (to my knowledge---I don't actually know what to look for).

**Comment by peastman** [comment]
Looks like that needs to be fixed too then!

I did mention this a couple of times when you were first setting up the builds: https://github.com/pandegroup/openmm/issues/513.  Perhaps it missed getting into the wiki though.  Sorry for the oversight.

**Comment by jchodera** [comment]
Sorry for missing this. Kyle probably added it at your suggestion but then
I removed it, not having read that part of the thread.

Entropy is our enemy.  Since there are so many people involved, it really
helps if we make it as easy as possible to find all the information we need
to do things (like use that wiki page to keep build environment specs up to
date).  We can discuss things in the issue tracker, but ultimately we need
to compile "the agreed on specification" somewhere unambiguous.

I will try to add the APP SDK back in, but how will this work with both
CUDA 7 and the APP SDK?  Do we compile CUDA against CUDA7 and OpenCL with
APP?

**Comment by peastman** [comment]
Correct.  The only things that need to reference the APP SDK are OPENCL_INCLUDE_DIR and OPENCL_LIBRARY.

**Comment by jchodera** [comment]
Great! Update the wiki when you get a chance.

**Comment by peastman** [comment]
Done.  I think everything on it is now correct for 6.3.

**Comment by peastman** [comment]
The person at Nvidia I'm corresponding with is only observing a small performance difference: 173 ns/day on Linux vs. 158 on Windows (164 if he enables TCC mode).  That's using the benchmark script in the examples directory:

    python benchmark.py --platform=CUDA --test=pme --seconds=10

In contrast, I get 178 under Linux and 80 under Windows.  That's on a Titan X.  Can anyone else try that and see what you get?

**Comment by jchodera** [comment]
Working on rebuilding the vagrant VM with the AMD APP SDK...

**Comment by peastman** [comment]
Sigh.

It turns out someone\* installed Folding@Home on the computer I was using for benchmarking.  Whenever I was booted into Windows, it was running in the background!  Once I turned it off, the speed went up to 158 ns/day, the same as what the person at Nvidia reported.

I now get only a small performance difference between Windows and Linux, presumably due to the less efficient driver model. @mpharrigan do you still observe a large performance difference?

\* someone = Yutong

**Comment by proteneer** [comment]
sorry :(

**Comment by peastman** [comment]
Hey, I didn't know you were monitoring this repo.  Hi there! :)

**Comment by informatorius** [comment]
I still can reproduce the issue with FahBench and OpenMM 6.3 latest on Windows. 
Your python benchmark.py script has no problem and also rmcgibbo's performance tracking does not show the issue. 
Maybe you can try it with FahBench yourself to see the issue?

**Comment by peastman** [comment]
I don't know much of anything about FahBench.  I've never looked at the code for it, and I don't think it's been updated in a long time.  I have no idea what it's doing.  But if all the regular benchmarks are fine, and only FahBench is having problems, that suggests pretty strongly the problem is in FahBench.

**Comment by mpharrigan** [comment]
I've been working on it recently. Could it be because we're using vs2013?

**Comment by peastman** [comment]
Seems a little unlikely, but I can't rule out the possibility.  Can you try compiling with 2010?

**Comment by mpharrigan** [comment]
I will do that when I'm at home (with my windows desktop)

**Comment by mpharrigan** [comment]
Tried with cuda. Updated table below

precision | platform | 6.2 | 6.3
-----------|----------------|-----|------
single | OpenCL | 23.4 | 2.0
single | CUDA | 24.4 | 5.3
double | OpenCL | 7.18 | error

**Comment by peastman** [comment]
Try running the CUDA version in nvvp?  Also, what does "error" mean?  What is the error?

**Comment by mpharrigan** [comment]
I emailed you the nvvp results

**Comment by mpharrigan** [comment]
CUDA + vs2010 gives 6 ns / day
OpenCL + vs2010 gives `no registered Platform called "OpenCL"` even though OpenCL is compiled and present in the plugin directory

**Comment by mpharrigan** [comment]
I no longer get an error with double precision. I've updated the table. I don't remember what the error was

**Comment by rmcgibbo** [comment]
Do you get the same speed just using the benchmarking script in examples/
vs. Fahbench?

**Comment by peastman** [comment]
From your profile, that looks like a very messed up simulation.  I'd guess that your coordinates are all nan.

The first thing I notice is that you're using more constraints than just waters and hydrogens.  I can tell that because it's using CCMA for constraints.  Is that intentional?  But the constraint algorithm never converges.  It runs 150 iterations every time before just giving up.  That indicates that either 1) your positions are nan, 2) your positions are so far from satisfying the constraints that it's outside the radius of convergence, or 3) you have a singular constraint matrix.  Then it spends a huge amount of time in what should be a trivial kernel, sorting a list of numbers.  That's another good indication that your coordinates are nan.

**Comment by mpharrigan** [comment]
Was there some incompatibility introduced in serialization? This system was probably serialized with 5.1 ...

**Comment by peastman** [comment]
There shouldn't have been.  The first thing to do is identify exactly where things are getting messed up.  Are the coordinates and velocities correct at the start of the simulation?  If not, something's going wrong with loading the initial state.  But if the initial values are OK, are they still correct after one time step?  If not, try looking at the initial forces.  You may find that some of them are huge, which would cause the simulation to immediately blow up.  Is that the case?  If so, are all of them bad, or just a few?  Try removing Force objects to pin down which one is responsible for the bad forces.

**Comment by mpharrigan** [comment]
red herrings all around with this issue. I think everything is fixed now with #1037
